File: iosApp/iosApp/Modules/ImprovementProposals/ImprovementProposalsPresenter.swift
// Created by web3d3v on 30/07/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation

enum ImprovementProposalsPresenterEvent {
    case didSelectCategory(idx: Int)
    case filterBy(text: String)
    case vote(idx: Int)
    case select(idx: Int)
    case dismiss
}

protocol ImprovementProposalsPresenter {

    func present()
    func handle(_ event: ImprovementProposalsPresenterEvent)
}

final class DefaultImprovementProposalsPresenter {
    private weak var view: ImprovementProposalsView?
    private let interactor: ImprovementProposalsInteractor
    private let wireframe: ImprovementProposalsWireframe

    private var proposals = [ImprovementProposal]()
    private var category: ImprovementProposal.Category = .infrastructure
    private var filterText: String = ""

    init(
        view: ImprovementProposalsView,
        interactor: ImprovementProposalsInteractor,
        wireframe: ImprovementProposalsWireframe
    ) {
        self.view = view
        self.interactor = interactor
        self.wireframe = wireframe
    }
}

extension DefaultImprovementProposalsPresenter: ImprovementProposalsPresenter {

    func present() {
        if proposals.isEmpty { view?.update(with: .loading) }
        interactor.fetchAllFeatures { [weak self] result in
            switch result {
            case let .success(proposals):
                self?.proposals = proposals
                self?.updateView()
            case let .failure(error):
                // TODO: Error to be property and just call update view
                self?.view?.update(
                    with: .error(
                        error: .init(
                            title: Localized("Error"),
                            body: error.localizedDescription,
                            actions: [
                                Localized("ok")
                            ]
                        )
                    )
                )
            }
        }
    }

    func handle(_ event: ImprovementProposalsPresenterEvent) {
        switch event {
        case let .didSelectCategory(idx):
            category = ImprovementProposal.Category.allCases[idx]
            updateView()
        case let .filterBy(text):
            filterText = text
            updateView()
        case let .select(idx):
            let proposals = currentProposals()
            wireframe.navigate(
                to: .proposal(proposal: proposals[idx], proposals: proposals)
            )
        case let .vote(idx):
            wireframe.navigate(to: .vote(proposal: currentProposals()[idx]))
        case .dismiss:
            wireframe.navigate(to: .dismiss)
        }
    }
}

private extension DefaultImprovementProposalsPresenter {
    
    func updateView() {
        view?.update(with: viewModel())
    }

    func viewModel() -> ImprovementProposalsViewModel {
        let proposals = currentProposals()
        let sectionType: ImprovementProposalsViewModel.Section.`Type` = .from(category)
        let section: ImprovementProposalsViewModel.Section = .init(
            title: sectionType.stringValue + " (\(proposals))",
            description: sectionType.descriptionValue,
            type: sectionType,
            items: proposals.compactMap { proposalViewModel(from: $0) },
            footer: nil
        )

        return .loaded(
            sections: [section],
            selectedSectionType: sectionType
        )
    }

    func proposalViewModel(from feature: ImprovementProposal) -> ImprovementProposalsViewModel.Item {
        .init(
            id: feature.id,
            title: feature.title,
            subtitle: proposalSubtitle(for: feature),
            buttonTitle: Localized("proposals.cell.button.vote")
        )
    }

    func proposalSubtitle(for feature: ImprovementProposal) -> String {
        feature.hashTag
            + "  |  "
            + Localized("proposals.cell.votes", arg: feature.votes.stringValue)
    }
}

private extension DefaultImprovementProposalsPresenter {

    func currentProposals() -> [ImprovementProposal] {
        proposals.filter { $0.category == category }
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/improvementProposals/ImprovementProposalsPresenter.kt
package com.sonsofcrypto.web3walletcore.modules.improvementProposals

import com.sonsofcrypto.web3lib.utils.WeakRef
import com.sonsofcrypto.web3lib.utils.bgDispatcher
import com.sonsofcrypto.web3lib.utils.withUICxt
import com.sonsofcrypto.web3walletcore.common.viewModels.ErrorViewModel
import com.sonsofcrypto.web3walletcore.common.viewModels.with
import com.sonsofcrypto.web3walletcore.extensions.Localized
import com.sonsofcrypto.web3walletcore.modules.improvementProposals.ImprovementProposalsViewModel.*
import com.sonsofcrypto.web3walletcore.modules.improvementProposals.ImprovementProposalsWireframeDestination.*
import com.sonsofcrypto.web3walletcore.services.improvementProposals.ImprovementProposal
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch

sealed class ImprovementProposalsPresenterEvent {
    /** Selected category changed */
    data class Category(val idx: Int): ImprovementProposalsPresenterEvent()
    /** Vote on proposal at index */
    data class Vote(val idx: Int): ImprovementProposalsPresenterEvent()
    /** Show details of proposal */
    data class Proposal(val idx: Int): ImprovementProposalsPresenterEvent()
    /** Did select alert action */
    data class AlertAction(val idx: Int): ImprovementProposalsPresenterEvent()
    /** Dismiss proposals module */
    object Dismiss: ImprovementProposalsPresenterEvent()
}

interface ImprovementProposalsPresenter {
    fun present()
    fun handle(event: ImprovementProposalsPresenterEvent)
}

class DefaultImprovementProposalsPresenter(
    private val view: WeakRef<ImprovementProposalsView>,
    private val wireframe: ImprovementProposalsWireframe,
    private val interactor: ImprovementProposalsInteractor,
) : ImprovementProposalsPresenter {
    private var proposals: List<ImprovementProposal> = emptyList()
    private var error: Throwable? = null
    private var selectedCategoryIdx: Int = 0
    private val scope = CoroutineScope(bgDispatcher)

    override fun present() {
        updateView()
        val errHandler = CoroutineExceptionHandler { _, err ->
            error = err
            updateView()
        }
        scope.launch(errHandler) {
            val result = interactor.fetchProposals()
            withUICxt {
                proposals = result
                updateView()
            }
        }
    }

    override fun handle(event: ImprovementProposalsPresenterEvent) {
        when (event) {
            is ImprovementProposalsPresenterEvent.Category -> {
                selectedCategoryIdx = event.idx
                updateView()
            }
            is ImprovementProposalsPresenterEvent.Vote -> wireframe.navigate(
                Vote(proposalAt(event.idx))
            )
            is ImprovementProposalsPresenterEvent.Proposal -> wireframe.navigate(
                Proposal(proposals(selectedCategoryIdx)[event.idx])
            )
            is ImprovementProposalsPresenterEvent.AlertAction -> {
                error = null
                updateView()
            }
            is ImprovementProposalsPresenterEvent.Dismiss -> wireframe.navigate(Dismiss)
        }
    }

    private fun updateView() {
        view.get()?.update(viewModel())
    }

    private fun viewModel(): ImprovementProposalsViewModel {
        if (error != null)
            return Error(ErrorViewModel.with(error!!))
        if (proposals.isEmpty())
            return Loading
        return Loaded(
            ImprovementProposal.Category.values().map { category ->
                Category(
                    category.title(),
                    category.description(),
                    proposals.filter { it.category == category }
                        .map { Item(it.id, it.title, proposalSubtitle(it)) }
                )
            },
            selectedCategoryIdx
        )
    }

    private fun proposalAt(idx: Int): ImprovementProposal {
        return proposals(selectedCategoryIdx)[idx]
    }

    private fun proposals(categoryIdx: Int): List<ImprovementProposal> {
        val category = ImprovementProposal.Category.values()[categoryIdx]
        return proposals.filter { it.category == category }
    }

    private fun proposalSubtitle(proposal: ImprovementProposal): String {
        return Localized("proposal.hashTag", proposal.id) + "  |  " +
            Localized("proposals.votes", proposal.votes)
    }
}
